OrbitLinuxTracing/ThreadStateVisitor.cpp

```diff
@@ -0,0 +1,272 @@
+      listener_->OnThreadStateSlice(std::move(out_slice.value()));
+    }
+  }
+  if (event->GetNextTid() != 0) {
+    std::optional<ThreadStateSlice> in_slice =
+        state_manager_.OnContextSwitchIn(event->GetTimestamp(), event->GetNextTid());
+    if (in_slice.has_value()) {
+      listener_->OnThreadStateSlice(std::move(in_slice.value()));
+    }
+  }
+}
+
+void ThreadStateVisitor::visit(SchedWakeupPerfEvent* event) {
+  CHECK(listener_ != nullptr);
+  std::optional<ThreadStateSlice> state_slice =
+      state_manager_.OnSchedWakeup(event->GetTimestamp(), event->GetWokenTid());
+  if (state_slice.has_value()) {
+    listener_->OnThreadStateSlice(std::move(state_slice.value()));
+  }
+}
+
+void ThreadStateVisitor::ProcessRemainingOpenStates(uint64_t timestamp_ns) {
+  CHECK(listener_ != nullptr);
+  std::vector<ThreadStateSlice> state_slices = state_manager_.OnCaptureFinished(timestamp_ns);
+  for (ThreadStateSlice& slice : state_slices) {
+    listener_->OnThreadStateSlice(slice);
+  }
+}
+
+// Associates a ThreadStateSlice::ThreadState to a thread state character retrieved from
+// /proc/<pid>/stat or the `ps` command. The possible characters are obtained from
+// /sys/kernel/debug/tracing/events/sched/sched_switch/format and compared with the ones listed in
+// https://man7.org/linux/man-pages/man5/proc.5.html and
+// https://www.man7.org/linux/man-pages/man1/ps.1.html#PROCESS_STATE_CODES to make sure we are not
+// missing any additional valid one.
+std::optional<ThreadStateSlice::ThreadState> ThreadStateVisitor::GetThreadStateFromChar(char c) {
+  switch (c) {
+    case 'R':
+      return ThreadStateSlice::kRunnable;
+    case 'S':
+      return ThreadStateSlice::kInterruptibleSleep;
+    case 'D':
+      return ThreadStateSlice::kUninterruptibleSleep;
+    case 'T':
+      return ThreadStateSlice::kStopped;
+    case 't':
+      return ThreadStateSlice::kTraced;
+    case 'X':
+      return ThreadStateSlice::kDead;
+    case 'Z':
+      return ThreadStateSlice::kZombie;
+    case 'P':
+      // Note that 'P' (Parked) is only valid from Linux 3.9 to 3.13, but we still include it as it
+      // is mentioned in /sys/kernel/debug/tracing/events/sched/sched_switch/format and in
+      // https://github.com/torvalds/linux/blob/master/fs/proc/array.c.
+      return ThreadStateSlice::kParked;
+    case 'I':
+      // 'I' (Idle) only applies to kernel threads. See
+      // https://github.com/torvalds/linux/commit/06eb61844d841d0032a9950ce7f8e783ee49c0d0.
+      return ThreadStateSlice::kIdle;
+    default:
+      return std::nullopt;
+  }
+}
+
+// Associates a ThreadStateSlice::ThreadState to the bits of the prev_state field of the
+// sched:sched_switch tracepoint. The association is given away by "print ftm" in
+// /sys/kernel/debug/tracing/events/sched/sched_switch/format or by
+// https://github.com/torvalds/linux/blob/master/fs/proc/array.c.
+ThreadStateSlice::ThreadState ThreadStateVisitor::GetThreadStateFromBits(uint64_t bits) {
```

Comment:
It feels a bit strange, that this method returns runnable in case of an invalid value, and the one above returns nullopt. I think the should behave the same. Also I think there should be a CHECK somewhere (either on the caller or on the callee), that the will be never an "unexpected" value.

Reply:
In which case, there would be invalid values again.

Reply:
> Can multiple bits of the last byte of "bits" be set to 1 at the same time?

Maybe, but I don't think I observed it when testing.

> What would that mean?

I don't know.

In https://github.com/torvalds/linux/blob/master/fs/proc/array.c the comment says 
"The task state array is a strange "bitmap" of reasons to sleep. Thus "running" is zero, and you can test for combinations of others with simple bit tests."
In such a situation, I prefer to pick the first state rather than none of them. On the other hand I didn't want to add complexity to support multiple states at once when it might only happen in some very specific situation. It's a compromise.
But I'll definitely add a log for it.

Reply:
cool, lgtm!